### PR TITLE
Prevent deleted user commenting and posting

### PR DESCRIPTION
**Pull request by Discordius**

### CODE REVIEW

**Comment by maaku** [comment]
I'm the user in question that triggered this PR, although not the only one who has been using the "deleted accounts post anonymously" trick, as I remember seeing at least a few other `[anonymous]` posts. Why block off this capability? It is actually quite useful. Official support for anonymous posting would of course be better, but in the meantime this works well enough.

File: packages/lesswrong/lib/collections/posts/collection.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ import { userIsPostGroupOrganizer } from './helpers';
 const options: MutationOptions<DbPost> = {
   newCheck: (user: DbUser|null) => {
     if (!user) return false;
+    if (user.deleted) return false;
     return userCanDo(user, 'posts.new')
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/posts/collection.ts:12]: This seems like the wrong place to enforce this. Isn't there a way in the user <-> posts.new relation that can run logic?

File: packages/lesswrong/lib/collections/users/helpers.ts [hunk 2]
```diff
@@ -164,6 +164,10 @@ export const userIsAllowedToComment = (user: UsersCurrent|DbUser|null, post: Pos
     return false
   }
 
+  if (user.deleted) {
+    return false
+  }
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/users/helpers.ts:169]: Same comment here, although here it at least looks like a this function is already being used for this. But you could you still imagine that this is an enforcement against all user new comments and being enforced elsewhere.

**Review by jpaddison3 — APPROVED** [review]
I feel like we should do this differently, but, assuming you've tested this, I'm fine with you thinking about whether you agree with my comments and making your own call.

**Comment by darkruby501** [comment]
This still needs testing. I said I'd do it.

File: packages/lesswrong/lib/collections/posts/collection.ts (hunk 1 above)
Comment (Discordius) [reply packages/lesswrong/lib/collections/posts/collection.ts:12]: Yeah, I don't actually really like the `userCanDo` paradigm, and would have preferred collections to generally just have generic permission functions. On the margin, I prefer doing things in readable code, as opposed to having permissions like this handled in some indirect way on the Collection object.

**Comment by Discordius** [comment]
It's definitely confusing for users to see comments by `[anonymous]` appearing, without any explanation for what is happening. And in general there are various reasons for why I am not a huge fan for supporting anonymous commenting without any caveats (I am excited for allowing high-karma users to make individual posts and comments anonymous, but in a way that is rate-limited, or somehow otherwise abuse-proof). If someone wants to be anonymous, the recommended way is to just create a new account. That way they also don't have to worry about their identity leaking in weird ways.

**Comment by Raemon** [comment]
I just tested this. It prevents people from making comments and posts. I'm not a fan of the current UI experience for posts (where it doesn't tell you you can't make a post until after you hit submit)

It also doesn't prevent deleted users from sending PMs (which is also confusing)

I think I prefer solving this by treating deleted users as "banned" in all respects, so we don't need to remember to manually remove every feature that we create from them.

**Comment by Raemon** [comment]
(That said, I think it is probably better to merge this current version and then worry about making future improvements later)

**Comment by Raemon** [comment]
I'm merging this for now. I may make further improvements in separate PR this week if it seems appropriate.

**Pull request merged**